File: swift/Sources/SignalClient/zkgroup/ClientZkReceiptOperations.swift
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class ClientZkReceiptOperations {

  let serverPublicParams: ServerPublicParams

  public init(serverPublicParams: ServerPublicParams) {
    self.serverPublicParams = serverPublicParams
  }

  public func createReceiptCredentialRequestContext(receiptSerial: ReceiptSerial) throws -> ReceiptCredentialRequestContext {
    return try createReceiptCredentialRequestContext(randomness: Randomness.generate(), receiptSerial: receiptSerial)
  }

  public func createReceiptCredentialRequestContext(randomness: Randomness, receiptSerial: ReceiptSerial) throws -> ReceiptCredentialRequestContext {
    return try serverPublicParams.withUnsafePointerToSerialized { serverPublicParams in
      try randomness.withUnsafePointerToBytes { randomness in
        try receiptSerial.withUnsafePointerToSerialized { receiptSerial in
          try invokeFnReturningSerialized {
            signal_server_public_params_create_receipt_credential_request_context_deterministic($0, serverPublicParams, randomness, receiptSerial)
          }
        }
      }
    }
  }

  public func receiveReceiptCredential(receiptCredentialRequestContext: ReceiptCredentialRequestContext, receiptCredentialResponse: ReceiptCredentialResponse) throws -> ReceiptCredential {
    return try serverPublicParams.withUnsafePointerToSerialized { serverPublicParams in
      try receiptCredentialRequestContext.withUnsafePointerToSerialized { requestContext in
        try receiptCredentialResponse.withUnsafePointerToSerialized { response in
          try invokeFnReturningSerialized {
            signal_server_public_params_receive_receipt_credential($0, serverPublicParams, requestContext, response)
          }
        }
      }
    }
  }

  public func createReceiptCredentialPresentation(receiptCredential: ReceiptCredential) throws -> ReceiptCredentialPresentation {
    return try createReceiptCredentialPresentation(randomness: Randomness.generate(), receiptCredential: receiptCredential)
  }

  public func createReceiptCredentialPresentation(randomness: Randomness, receiptCredential: ReceiptCredential) throws -> ReceiptCredentialPresentation {
    return try serverPublicParams.withUnsafePointerToSerialized { serverPublicParams in
      try randomness.withUnsafePointerToBytes { randomness in
        try receiptCredential.withUnsafePointerToSerialized { receiptCredential in
          try invokeFnReturningSerialized {
            signal_server_public_params_create_receipt_credential_presentation_deterministic($0, serverPublicParams, randomness, receiptCredential)
          }
        }
      }
    }
  }

}

File: java/shared/java/org/signal/libsignal/zkgroup/receipts/ClientZkReceiptOperations.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.receipts;

import static org.signal.libsignal.internal.FilterExceptions.filterExceptions;
import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.ServerPublicParams;
import org.signal.libsignal.zkgroup.VerificationFailedException;

public class ClientZkReceiptOperations {

  private final ServerPublicParams serverPublicParams;

  public ClientZkReceiptOperations(ServerPublicParams serverPublicParams) {
    this.serverPublicParams = serverPublicParams;
  }

  public ReceiptCredentialRequestContext createReceiptCredentialRequestContext(
      ReceiptSerial receiptSerial) throws VerificationFailedException {
    return createReceiptCredentialRequestContext(new SecureRandom(), receiptSerial);
  }

  public ReceiptCredentialRequestContext createReceiptCredentialRequestContext(
      SecureRandom secureRandom, ReceiptSerial receiptSerial) throws VerificationFailedException {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        serverPublicParams.guardedMap(
            (serverPublicParams) ->
                Native.ServerPublicParams_CreateReceiptCredentialRequestContextDeterministic(
                    serverPublicParams, random, receiptSerial.getInternalContentsForJNI()));

    try {
      return new ReceiptCredentialRequestContext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ReceiptCredential receiveReceiptCredential(
      ReceiptCredentialRequestContext receiptCredentialRequestContext,
      ReceiptCredentialResponse receiptCredentialResponse)
      throws VerificationFailedException {
    byte[] newContents =
        filterExceptions(
            VerificationFailedException.class,
            () ->
                serverPublicParams.guardedMapChecked(
                    (publicParams) ->
                        Native.ServerPublicParams_ReceiveReceiptCredential(
                            publicParams,
                            receiptCredentialRequestContext.getInternalContentsForJNI(),
                            receiptCredentialResponse.getInternalContentsForJNI())));
    try {
      return new ReceiptCredential(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ReceiptCredentialPresentation createReceiptCredentialPresentation(
      ReceiptCredential receiptCredential) throws VerificationFailedException {
    return createReceiptCredentialPresentation(new SecureRandom(), receiptCredential);
  }

  public ReceiptCredentialPresentation createReceiptCredentialPresentation(
      SecureRandom secureRandom, ReceiptCredential receiptCredential)
      throws VerificationFailedException {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        serverPublicParams.guardedMap(
            (publicParams) ->
                Native.ServerPublicParams_CreateReceiptCredentialPresentationDeterministic(
                    publicParams, random, receiptCredential.getInternalContentsForJNI()));
    try {
      return new ReceiptCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}
